goshopify.go

```diff
@@ -308,17 +297,40 @@ func (c *Client) Do(req *http.Request, v interface{}) error {
 
 // doGetHeaders executes a request, decoding the response into `v` and also returns any response headers.
 func (c *Client) doGetHeaders(req *http.Request, v interface{}) (http.Header, error) {
-	resp, err := c.Client.Do(req)
-	if err != nil {
-		return nil, err
-	}
-	defer resp.Body.Close()
+	var resp *http.Response
+	var err error
 
-	err = CheckResponseError(resp)
-	if err != nil {
-		return nil, err
+	r := c.retries
+	for {
+		resp, err = c.Client.Do(req)
+		if err != nil {
+			return nil, err
+		}
+
+		err = CheckResponseError(resp)
+		if err != nil {
+			if e, isRetryErr := err.(RateLimitError); isRetryErr {
+				err = resp.Body.Close()
+				if err != nil {
+					return nil, err
+				}
+
+				// back off and retry
+				if r <= 0 {
+					return nil, err
+				}
+				wait := time.Duration(e.RetryAfter) * time.Second
+				time.Sleep(wait)
+				r--
+				continue
+			}
+			return nil, err
+		}
+		break
 	}
 
+	defer resp.Body.Close()
```

Comment:
This might still be problematic.
If you get e.g. a 429 and you do a retry then you won't do a close ont he body of that 429 request and leak that resource. You'd only do the defer Close() on the last request which succeeded.
I *think* the defer close needs to go in line 309 (and I think it should be ok with `resp` being bound at defer time)

Reply:
you dont want to do a defer in a loop tho? there's a close in the retry at line 313 to keep the response closed during the loop